### ISSUE TITLE
Yolov8 OBB validation bug

### ISSUE THREAD

**Issue by SirojbekSafarov** [ultralytics/ultralytics#7401]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

1) In ultralytics/engine/model.py val function 'custom = {'rect': True} ' custom hard coded. If the model is trained with rect=False, it will affect visualization and mAP.
<img width="728" alt="image" src="https://github.com/ultralytics/ultralytics/assets/45415334/9c6cd247-6aca-49a3-a49c-8e4858a3a74d">

2) In ultralytics/models/yolo/obb/val.py  '_process_batch' function iou is calculated like
'' iou = batch_probiou(gt_bboxes, torch.cat([detections[:, :4], detections[:, -2:-1]], dim=-1)) ''
predicted detection is like [x,y,h,w, class, angle]
but this 'torch.cat([detections[:, :4], detections[:, -2:-1]], dim=-1)' makes predicted bbox like [x,y,h,w, class] and it calculates IOU wrong.
<img width="624" alt="image" src="https://github.com/ultralytics/ultralytics/assets/45415334/f13ac0cd-f636-4d8e-b5e3-578f5cada746">

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Laughing-q** [comment]
@SirojbekSafarov we have `rect=True` by default and you actually could explicitly pass `rect=False` to make sure the inference images are square.
For the iou calculation, don't worry it's correct, the last element of `detection` is the angle dimension. :)

**Comment by SirojbekSafarov** [comment]
Actually I debuged it and this 'detections[:, -2:-1]'  does not get last element of the tensor it gets one before last one and that one is class number. If you have time try to debug it you can see :).

**Comment by Laughing-q** [comment]
@SirojbekSafarov oh right I think you are correct! The internal metric seems fine even this `detections[:, -2:-1]` is not as I expected so I didn't notice. Thanks! I'll made a PR to fix it.
BTW don't worry this won't affect the mAP in our table as these are evaluated by official DOTA website, and the models are fine. :)

**Comment by Laughing-q** [comment]
@SirojbekSafarov I just found the nms part might also be incorrect, I'll test it and probably re-evaluate our models. Hopefully we'll get better results! THANKS for the finding!!!

**Comment by Laughing-q** [comment]
@SirojbekSafarov ok I've fixed this issue by https://github.com/ultralytics/ultralytics/pull/7404.
And also I re-evaluated our obb models, the mAP stays the same, no worries about that. :)
